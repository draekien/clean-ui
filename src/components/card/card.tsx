// eslint-disable-next-line tsdoc/syntax
/** @jsxImportSource theme-ui */
import React from 'react';
import { Scrollable } from '../..';
import {
  cardContainerCss,
  CardContainerCssProps,
  cardFooterCss,
  cardHeaderCss,
} from './card.styles';

export interface CardProps extends CardContainerCssProps {
  /** element to render in a dedicated header div */
  header?: React.ReactNode;
  /** element to render in a dedicated footer div */
  footer?: React.ReactNode;
}

export const Card: React.FC<CardProps> = ({
  width = '20em',
  maxHeight = '',
  fullWidth = false,
  backgroundColor = 'background-lighter',
  borderColor = 'border',
  variant = 'medium',
  feature = false,
  noShadow = false,
  noPadding = false,
  header,
  footer,
  children,
  ...rest
}) => {
  return (
    <div
      {...rest}
      sx={cardContainerCss({
        width,
        maxHeight,
        fullWidth,
        backgroundColor,
        borderColor,
        variant,
        feature,
        noShadow,
        noPadding,
      })}>
      <Scrollable maxHeight={maxHeight}>
        {header && <div sx={cardHeaderCss(variant)}>{header}</div>}
        {children && <div>{children}</div>}
        {footer && <div sx={cardFooterCss(variant)}>{footer}</div>}
      </Scrollable>
    </div>
  );
};
