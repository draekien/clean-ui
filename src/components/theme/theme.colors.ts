export const colors = {
  text: '#292929',
  background: '#F7F7F7',
  primary: '#EF802F',
  secondary: '#DB5192',
  muted: '#D8D8D8',

  'p-000': '#FCE6D5',
  'p-100': '#F9CCAC',
  'p-200': '#F5B382',
  'p-300': '#F29959',
  'p-400': '#EF802F',

  's-000': '#F8DCE9',
  's-100': '#F1B9D3',
  's-200': '#E997BE',
  's-300': '#E274A8',
  's-400': '#DB5192',

  'b-000': '#F7F7F7',
  'b-100': '#EFEFEF',
  'b-200': '#D8D8D8',
  'b-300': '#B1B1B1',
  'b-400': '#898989',
  'b-500': '#626262',
  'b-600': '#3B3B3B',
  'b-700': '#292929',

  'info-000': '#CDE1EB',
  'info-400': '#066899',

  'success-000': '#D5EADC',
  'success-400': '#2E9852',

  'warning-000': '#FBDBD1',
  'warning-400': '#EA4318',

  'error-000': '#F9CCD4',
  'error-400': '#E00027',

  modes: {
    dark: {
      text: '#F7F7F7',
      background: '#292929',
      primary: '#EF802F',
      secondary: '#DB5192',
      muted: '#3B3B3B',
    },
  },
};
