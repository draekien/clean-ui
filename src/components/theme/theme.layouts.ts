export const layouts = {
  radii: {
    xs: '0.125rem',
    sm: '0.25rem',
    md: '0.5rem',
    lg: '0.75rem',
    xl: '1rem',
    max: 999,
  },
  space: {
    xxs: '0.25rem',
    xs: '0.5rem',
    sm: '0.75rem',
    md: '1rem',
    lg: '1.5rem',
    xl: '2rem',
    xxl: '2.5rem',
  },
  shadows: {
    sm: '0px 4px 4px rgba(61, 61, 61, 0.25), 0px 2px 4px -2px rgba(61, 61, 61, 0.25);',
    md: '0px 4px 8px rgba(61, 61, 61, 0.25), 0px 4px 8px -4px rgba(61, 61, 61, 0.25);',
    lg: '0px 4px 8px rgba(61, 61, 61, 0.25), 0px 16px 32px -16px rgba(61, 61, 61, 0.25)',
    text: '0px 2px 4px rgba(61, 61, 61, 0.25);',
    disabled: '0px 2px 4px -2px rgba(4, 40, 52, 0.25);',
  },
  zIndices: {
    lowest: 0,
    lower: 100,
    low: 200,
    middle: 300,
    high: 400,
    higher: 500,
    highest: 600,
    pagination: 1,
    topMenu: 200,
    dropdown: 300,
    drawer: 400,
    modal: 500,
    toast: 600,
    overlay: 600,
  },
};
