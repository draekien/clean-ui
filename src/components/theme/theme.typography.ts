export const typography = {
  fonts: {
    body: '"Barlow", sans-serif',
    mono: '"Fira Code", monospace',
  },
  fontSizes: {
    hero: '6rem',
    h1: '3rem',
    h2: '2.5rem',
    h3: '2rem',
    h4: '1.5rem',
    subtitle: '1rem',
    body: '1rem',
    eyebrow: '0.75rem',
    small: '0.75rem',
  },
  fontWeights: {
    thin: 300,
    regular: 400,
    medium: 500,
    semibold: 600,
    bold: 700,
  },
  lineHeights: {
    body: 1.5,
    heading: 1.125,
  },
};
