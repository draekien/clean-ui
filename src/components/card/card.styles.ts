import { Theme, ThemeUIStyleObject } from '@theme-ui/css';
import { alpha } from '@theme-ui/color';
import { Size } from '../../types/layouts';
import { abbreviate } from '../../utils/size-converter.util';
import { colors } from '../theme/theme.colors';

export interface CardContainerCssProps {
  /** should the card take up the full width of the parent container?
   *  @default false
   */
  fullWidth?: boolean;
  /** the width of the card. This is overriden if fullWidth is true
   *  @default '20em'
   */
  width?: string | number | string[] | number[];
  maxHeight?: string | number;
  /** the background color of the card
   *  @default 'background-lighter'
   */
  backgroundColor?: keyof typeof colors;
  /** the border color of the card
   *  @default 'border'
   */
  borderColor?: keyof typeof colors;
  /** the layout variants of the card. This determins the amount of
   *  padding, box shadow, and border radius on the card
   *  @default 'medium'
   */
  variant?: Size;
  /** should the card be rendered with a gradient background?
   *  @default false
   */
  feature?: boolean;
  /** should the card render without a dropshadow? */
  noShadow?: boolean;
  /** should the card render without padding? */
  noPadding?: boolean;
}

export const cardContainerCss = ({
  fullWidth,
  width,
  maxHeight,
  backgroundColor,
  borderColor,
  variant,
  feature,
  noShadow,
  noPadding,
}: CardContainerCssProps): ThemeUIStyleObject => {
  const size = abbreviate(variant || 'medium');
  const css: ThemeUIStyleObject = {
    display: fullWidth ? 'flex' : 'inline-flex',
    flexFlow: 'column',
    width: fullWidth ? '100%' : width,
    maxHeight: !maxHeight ? 'initial' : maxHeight,
    boxShadow: noShadow ? 'none' : size,
    padding: size,
    borderRadius: size,
  };

  if (feature) {
    css.background = (t: Theme) =>
      `linear-gradient(to right, ${alpha('p-400', 1)(t)} 0%, ${alpha(
        's-400',
        1
      )(t)} 51%, ${alpha('p-400', 1)(t)} 100%)`;
    css.color = 'b-000';
    css.backgroundSize = '200% auto';
  } else {
    css.backgroundColor = backgroundColor;
  }

  if (noShadow) {
    css.border = '1px solid';
    css.borderColor = borderColor;
  }

  if (noPadding) {
    css.padding = 0;
  }

  return css;
};

export const cardHeaderCss = (variant: Size): ThemeUIStyleObject => {
  const size = abbreviate(variant);
  const css: ThemeUIStyleObject = {
    pb: size,
  };

  return css;
};

export const cardFooterCss = (variant: Size): ThemeUIStyleObject => {
  const size = abbreviate(variant);
  const css: ThemeUIStyleObject = {
    pt: size,
  };

  return css;
};
