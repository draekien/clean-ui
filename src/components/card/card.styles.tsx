import { Theme, ThemeUIStyleObject } from '@theme-ui/css';
import { alpha } from '@theme-ui/color';
import { Size } from '../../types/layouts';
import { abbreviate } from '../../utils/size-converter.util';
import { colors } from '../theme/theme.colors';

export interface CardContainerCssProps {
  /** should the card take up the full width of the parent container?
   *  @default false
   */
  fullWidth?: boolean;
  /** the width of the card. This is overriden if fullWidth is true
   *  @default '20em'
   */
  width?: string | number | string[] | number[];
  /** the background color of the card
   *  @default 'background-lighter'
   */
  backgroundColor?: keyof typeof colors;
  /** the layout variants of the card. This determins the amount of
   *  padding, box shadow, and border radius on the card
   *  @default 'medium'
   */
  variant?: Size;
  /** should the card be rendered with a gradient background?
   *  @default false
   */
  feature?: boolean;
  /** should the card render without a dropshadow? */
  noShadow?: boolean;
}

export const cardContainerCss = ({
  fullWidth,
  width,
  backgroundColor,
  variant,
  feature,
  noShadow,
}: CardContainerCssProps): ThemeUIStyleObject => {
  const size = abbreviate(variant || 'medium');
  const css: ThemeUIStyleObject = {
    display: fullWidth ? 'flex' : 'inline-flex',
    flexFlow: 'column',
    width: fullWidth ? '100%' : width,
    boxShadow: noShadow ? 'none' : size,
    padding: size,
    borderRadius: size,
  };

  if (feature) {
    css.background = (t: Theme) =>
      `linear-gradient(to right, ${alpha('p-400', 1)(t)} 0%, ${alpha(
        's-400',
        1
      )(t)} 51%, ${alpha('p-400', 1)(t)} 100%)`;
    css.color = 'b-000';
    css.backgroundSize = '200% auto';
  } else {
    css.backgroundColor = backgroundColor;
  }

  if (noShadow) {
    css.border = '1px solid';
    css.borderColor = 'border';
  }

  return css;
};

export const cardHeaderCss = (variant: Size): ThemeUIStyleObject => {
  const size = abbreviate(variant);
  const css: ThemeUIStyleObject = {
    pb: size,
  };

  return css;
};

export const cardFooterCss = (variant: Size): ThemeUIStyleObject => {
  const size = abbreviate(variant);
  const css: ThemeUIStyleObject = {
    pt: size,
  };

  return css;
};
