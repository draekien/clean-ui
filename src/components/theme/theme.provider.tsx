/** @jsxImportSource theme-ui */
import React, { useEffect } from 'react';
import { ThemeProvider as ThemeUiProvider } from 'theme-ui';
import { CleanUiTheme, Theme } from './theme';

export interface CleanUiProps {
  theme?: Partial<Theme>;
  children: React.ReactNode;
}

const mergeThemes = (target: any, source: any) => {
  Object.keys(source).forEach((key) => {
    if (key === '__proto__' || key === 'constructor') return;

    if (source[key] && typeof source[key] === 'object') {
      mergeThemes((target[key] = target[key] || {}), source[key]);
      return;
    }

    target[key] = source[key];
  });
};

const addThemeFonts = () => {
  const fonts: { [key: string]: string } = {
    text:
      'https://fonts.googleapis.com/css2?family=Barlow:wght@200;300;400;500;600;700&family=Fira+Code&display=swap',
    icons:
      'https://fonts.googleapis.com/icon?family=Material+Icons|Material+Icons+Outlined|Material+Icons+Two+Tone|Material+Icons+Round|Material+Icons+Sharp',
  };

  const preconnectLink = document.createElement('link');
  preconnectLink.rel = 'preconnect';
  preconnectLink.href = 'https://fonts.gstatic.com';
  document.head.appendChild(preconnectLink);

  Object.keys(fonts).forEach((key) => {
    const linkExists = document.getElementById(key);

    if (!linkExists) {
      const link = document.createElement('link');
      link.id = key;
      link.rel = 'stylesheet';
      link.href = fonts[key];
      link.crossOrigin = 'anonymous';
      document.head.appendChild(link);
    }
  });
};

export const CleanUi = ({ theme, children }: CleanUiProps) => {
  const newTheme = JSON.parse(JSON.stringify(CleanUiTheme));

  if (theme) mergeThemes(newTheme, theme);

  useEffect(() => {
    addThemeFonts();
  }, []);

  return <ThemeUiProvider theme={newTheme}>{children}</ThemeUiProvider>;
};
