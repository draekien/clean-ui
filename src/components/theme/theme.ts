import { Theme as ThemeUiTheme } from '@theme-ui/css';
import { colors } from './theme.colors';
import { layouts } from './theme.layouts';
import { styles } from './theme.styles';
import { typography } from './theme.typography';
import { variants } from './theme.variants';

export type Theme = Partial<{}>;

export const CleanUiTheme: ThemeUiTheme = {
  useCustomProperties: false,
  useBorderBox: true,
  useRootStyles: true,
  useLocalStorage: true,
  breakpoints: ['40em', '56em', '64em'],
  colors,
  styles,
  ...variants,
  ...layouts,
  ...typography,
};
