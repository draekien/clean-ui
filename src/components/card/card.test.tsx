import React from 'react';
import { hexToRgb, themedRender } from '../../helpers/testHelpers';
import { abbreviate } from '../../utils/size-converter.util';
import { layouts } from '../theme/theme.layouts';
import { Size } from '../../types/layouts';
import { Card } from './card';
import { colors } from '../theme/theme.colors';

describe('Card Component', () => {
  it('Should match snapshot', () => {
    const { container } = themedRender(<Card>Card</Card>);

    expect(container).toMatchSnapshot();
  });

  it('Should render header and footer when provided', () => {
    const { container } = themedRender(
      <Card header="header" footer="footer">
        Card
      </Card>
    );

    expect(container.innerHTML).toContain('header');
    expect(container.innerHTML).toContain('footer');
  });

  it('Should render with full width when full width prop is true', () => {
    const { container } = themedRender(<Card fullWidth>Card</Card>);

    const styles = window.getComputedStyle(container.querySelector('div') as Element);
    expect(styles.display).toBe('flex');
    expect(styles.width).toBe('100%');
  });

  it('Should render with feature styles when feature prop is true', () => {
    const { container } = themedRender(<Card feature>Card</Card>);

    const styles = window.getComputedStyle(container.querySelector('div') as Element);
    expect(styles.backgroundSize).toBe('200% auto');
  });

  const sizes: Size[] = ['small', 'medium', 'large'];

  sizes.forEach((size) => {
    it(`Should render ${size} variant correctly`, () => {
      const { container } = themedRender(<Card variant={size}>Card</Card>);

      const styles = window.getComputedStyle(container.querySelector('div') as Element);
      expect(styles.padding).toBe(layouts.space[abbreviate(size)]);
      expect(styles.borderRadius).toBe(layouts.radii[abbreviate(size)]);
    });
  });

  it('Should render with provided width', () => {
    const { container } = themedRender(<Card width="20em">Card</Card>);

    const styles = window.getComputedStyle(container.querySelector('div') as Element);
    expect(styles.width).toBe('20em');
  });

  it('Should render with provided background color', () => {
    const { container } = themedRender(<Card backgroundColor="primary">Card</Card>);

    const styles = window.getComputedStyle(container.querySelector('div') as Element);
    expect(styles.backgroundColor).toBe(hexToRgb(colors.primary));
  });

  it('Should render with border when noShadow is true', () => {
    const { container } = themedRender(<Card noShadow>Card</Card>);

    const styles = window.getComputedStyle(container.querySelector('div') as Element);
    expect(styles.boxShadow).toBe('none');
    expect(styles.border).toBe('1px solid');
    expect(styles.borderColor).toBe('#b1b1b1');
  });

  it('Should render with the correct border color', () => {
    const { container } = themedRender(
      <Card borderColor="primary" noShadow>
        Card
      </Card>
    );

    const styles = window.getComputedStyle(container.querySelector('div') as Element);
    expect(styles.borderColor).toBe('#ef802f');
  });
});
