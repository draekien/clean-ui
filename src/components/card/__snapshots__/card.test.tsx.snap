// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component Should match snapshot 1`] = `
<div>
  <div
    class="css-1xkb9fc-Card"
  >
    <div
      style="position: relative; overflow: hidden; width: 100%; height: auto; min-height: 0;"
    >
      <div
        style="position: relative; overflow: scroll; margin-right: 0px; margin-bottom: 0px; min-height: 0; max-height: calc( + 0px);"
      >
        <div>
          Card
        </div>
      </div>
      <div
        style="position: absolute; height: 6px; display: none; right: 2px; bottom: 2px; left: 2px; border-radius: 3px;"
      >
        <div
          style="position: relative; display: block; height: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
        />
      </div>
      <div
        style="position: absolute; width: 6px; display: none; right: 2px; bottom: 2px; top: 2px; border-radius: 3px;"
      >
        <div
          style="position: relative; display: block; width: 100%; cursor: pointer; border-radius: inherit; background-color: rgba(0, 0, 0, 0.2);"
        />
      </div>
    </div>
  </div>
</div>
`;
